test/validation/api/scheduler/scheduler.c

```diff
@@ -2484,6 +2489,120 @@ static void scheduler_test_ordered_and_plain(void)
 	scheduler_test_sched_and_plain(ODP_SCHED_SYNC_ORDERED);
 }
 
+static int atomicity_test_run(void *arg)
+{
+	thread_args_t *args = (thread_args_t *)arg;
+	odp_event_t ev;
+	odp_queue_t atomic_queue = args->globals->atomicity_q.handle;
+	odp_queue_t from;
+	odp_atomic_u32_t *state;
+	uint32_t old;
+	uint32_t num_processed = 0;
+
+	if (args->num_workers > 1)
+		odp_barrier_wait(&globals->barrier);
+
+	while (num_processed < ATOMICITY_ROUNDS) {
+		ev  = odp_schedule(&from, ODP_SCHED_NO_WAIT);
+		if (ev == ODP_EVENT_INVALID)
+			continue;
+
+		CU_ASSERT(from == atomic_queue);
+		if (from != atomic_queue) {
+			odp_event_free(ev);
+			continue;
+		}
+
+		state = odp_queue_context(from);
+		CU_ASSERT_FATAL(state != NULL);
+
+		old = 0;
+		CU_ASSERT_FATAL(odp_atomic_cas_acq_rel_u32(state, &old, 1));
+
+		/* Hold atomic context a while to better reveal possible atomicity bugs */
+		odp_time_wait_ns(ODP_TIME_MSEC_IN_NS);
+
+		old = 1;
+		CU_ASSERT_FATAL(odp_atomic_cas_acq_rel_u32(state, &old, 0));
+
+		CU_ASSERT_FATAL(odp_queue_enq(from, ev) == 0);
+
+		num_processed++;
+	}
+
+	odp_schedule_pause();
+	while ((ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT)) != ODP_EVENT_INVALID)
+		CU_ASSERT_FATAL(odp_queue_enq(atomic_queue, ev) == 0);
```

Comment:
I was about to write a comment here that this loop here seems redundant because of the draining that is done later, but then I realized that because we are using a non-parallel queue we have to pause the scheduler and get rid of possible prefetched events and our scheduling context in order to not block other threads and end up in a deadlock with the barrier. A short comment in the code might be helpful if this is not something that should be obvious anyway. 

Otherwise this commit looks good to me.

Reply:
Okay, I'll add a comment in V2.